modules/dlsobackend/dlsobackend.cc

```diff
@@ -0,0 +1,722 @@
+{
+  if (api->list == nullptr)
+    return false;
+
+  string qname = target.toString();
+  bool success;
+
+  success = api->list(api->handle, qname.size(), qname.c_str(), domain_id);
+
+  if (!success)
+    throw PDNSException("Backend failed");
+
+  return success;
+}
+
+void fill_cb(const void* ptr, const struct resource_record* record)
+{
+  DNSResourceRecord* rr = (DNSResourceRecord*)ptr;
+  rr->qtype = record->qtype;
+  string qname = string(record->qname, record->qname_len);
+  rr->qname = DNSName(qname);
+  rr->qclass = QClass::IN;
+  rr->content = string(record->content, record->content_len);
+  rr->ttl = record->ttl;
+  rr->auth = record->auth;
+  rr->scopeMask = record->scope_mask;
+  rr->domain_id = record->domain_id;
+}
+
+bool DlsoBackend::get(DNSResourceRecord& rr)
+{
+  bool success = api->get(api->handle, fill_cb, &rr);
+
+  if (!success) {
+    in_query = false;
+    return false;
+  }
+
+  return true;
+}
+
+void fill_meta_cb(const void* ptr, uint8_t value_len, const struct dns_value* values)
+{
+  std::vector<std::string>* meta = (std::vector<std::string>*)ptr;
+  for (uint8_t i = 0; i < value_len; i++) {
+    const struct dns_value* value = &values[i];
+    string value_s = string(value->value, value->value_len);
+
+    (*meta).push_back(value_s);
+  }
+}
+
+void fill_metas_cb(const void* ptr, uint8_t meta_len, const struct dns_meta* c_metas)
+{
+  std::map<std::string, std::vector<std::string>>* metas = (std::map<std::string, std::vector<std::string>>*)ptr;
+  for (uint8_t i = 0; i < meta_len; i++) {
+    const struct dns_meta* meta = &c_metas[i];
+    string property = string(meta->property, meta->property_len);
+    const struct dns_value* values = meta->values;
+
+    fill_meta_cb(&(*metas)[property], meta->value_len, values);
+  }
+}
+
+bool DlsoBackend::getAllDomainMetadata(const DNSName& name, std::map<std::string, std::vector<std::string>>& metas)
+{
+  if (d_dnssec == false)
+    return false;
+  if (api->get_metas == nullptr)
+    return false;
+
+  metas.clear();
+
+  string qname = name.toString();
+  return api->get_metas(api->handle, qname.size(), qname.c_str(), fill_metas_cb, &metas);
+}
+
+bool DlsoBackend::getDomainMetadata(const DNSName& name, const std::string& kind, std::vector<std::string>& meta)
+{
+  if (api->get_meta == nullptr)
+    return false;
+
+  meta.clear();
+
+  string qname = name.toString();
+  return api->get_meta(api->handle, qname.size(), qname.c_str(), kind.size(), kind.c_str(), fill_meta_cb, &meta);
+}
+
+bool DlsoBackend::setDomainMetadata(const DNSName& name, const std::string& kind, const std::vector<std::string>& meta)
+{
+  if (api->set_meta == nullptr)
+    return false;
+
+  struct dns_value* values = (struct dns_value*)calloc(meta.size(), sizeof(struct dns_value));
+  if (values == nullptr)
+    return false;
+  for (uint i = 0; i < meta.size(); i++) {
+    values[i].value_len = meta[i].size();
+    values[i].value = meta[i].c_str();
+  }
```

Comment:
Same maybe it's a good spot for a vector and `vector::data()`, but same concern with placement (is the data directly constructed on heap).

Reply:
Please use a `vector` :)